sys/net/gnrc/link_layer/lorawan/gnrc_lorawan_mlme.c

```diff
@@ -0,0 +1,344 @@
+                mac->mlme.backoff_budget = GNRC_LORAWAN_BACKOFF_BUDGET_3;
+                break;
+        }
+
+    }
+
+    counter--;
+    mac->mlme.backoff_state = state << 5 | (counter & 0x1F);
+    xtimer_set_msg(&mac->mlme.backoff_timer,
+                   GNRC_LORAWAN_BACKOFF_WINDOW_TICK,
+                   &mac->mlme.backoff_msg, thread_getpid());
+
+}
+
+static void _mlme_set(gnrc_lorawan_t *mac, const mlme_request_t *mlme_request,
+                               mlme_confirm_t *mlme_confirm)
+{
+    mlme_confirm->status = -EINVAL;
+    switch(mlme_request->mib.type) {
+        case MIB_ACTIVATION_METHOD:
+            if(mlme_request->mib.activation != MLME_ACTIVATION_OTAA) {
+                mlme_confirm->status = GNRC_LORAWAN_REQ_STATUS_SUCCESS;
+                mac->mlme.activation = mlme_request->mib.activation;
+            }
+            break;
+        default:
+            break;
+    }
+}
+
+static void _mlme_get(gnrc_lorawan_t *mac, const mlme_request_t *mlme_request,
+                               mlme_confirm_t *mlme_confirm)
+{
+    switch(mlme_request->mib.type) {
+        case MIB_ACTIVATION_METHOD:
+            mlme_confirm->status = GNRC_LORAWAN_REQ_STATUS_SUCCESS;
+            mlme_confirm->mib.activation = mac->mlme.activation;
+            break;
+        default:
+            mlme_confirm->status = -EINVAL;
+            break;
+    }
+}
+
+void gnrc_lorawan_mlme_request(gnrc_lorawan_t *mac, const mlme_request_t *mlme_request,
+                               mlme_confirm_t *mlme_confirm)
+{
+    switch (mlme_request->type) {
+        case MLME_JOIN:
+            if(mac->mlme.activation != MLME_ACTIVATION_NONE) {
+                mlme_confirm->status = -EINVAL;
+                break;
+            }
+            if (!gnrc_lorawan_mac_acquire(mac)) {
+                mlme_confirm->status = -EBUSY;
+                break;
+            }
+
+            if (mac->mlme.backoff_budget < 0) {
+                mlme_confirm->status = -EDQUOT;
+                break;
+            }
+            memcpy(mac->appskey, mlme_request->join.appkey, LORAMAC_APPKEY_LEN);
+            mlme_confirm->status = gnrc_lorawan_send_join_request(mac, mlme_request->join.deveui,
+                                                                  mlme_request->join.appeui, mlme_request->join.appkey, mlme_request->join.dr);
+            break;
+        case MLME_LINK_CHECK:
+            mac->mlme.pending_mlme_opts |= GNRC_LORAWAN_MLME_OPTS_LINK_CHECK_REQ;
+            mlme_confirm->status = GNRC_LORAWAN_REQ_STATUS_DEFERRED;
+            break;
+        case MLME_SET:
+            _mlme_set(mac, mlme_request, mlme_confirm);
+            break;
+        case MLME_GET:
+            _mlme_get(mac, mlme_request, mlme_confirm);
+            break;
+        case MLME_RESET:
+            gnrc_lorawan_reset(mac);
+            mlme_confirm->status = GNRC_LORAWAN_REQ_STATUS_SUCCESS;
+            break;
+        default:
+            break;
```

Comment:
mhm, maybe you use `break` more often in `default` then I thought 😄

Reply:
addressed :) (see d.g https://github.com/RIOT-OS/RIOT/pull/11022/commits/bda25749eb84ba2310409f887164eb7911ebc4e9)